File: src-tauri/builtin-plugins/sdk/sdk.js
/**
 * @agentjax/sdk — Shared utilities for AgentJax provider plugins.
 *
 * This module is pre-registered via StaticModuleLoader so all built-in
 * and user plugins can import shared helpers without bundling duplicates.
 *
 * All functions are pure — no mutable state, no side effects.
 */

/**
 * Append query parameters to a URL, preserving any existing query string.
 * @param {string} url
 * @param {Record<string,string>} [queryParams]
 * @returns {string}
 */
export function withQuery(url, queryParams) {
  const pairs = Object.entries(queryParams || {})
    .filter(([key, value]) => String(key).trim() && String(value).trim())
    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`);
  if (!pairs.length) return url;
  const separator = url.includes("?") ? (url.endsWith("?") || url.endsWith("&") ? "" : "&") : "?";
  return `${url}${separator}${pairs.join("&")}`;
}

/**
 * Build a normalized stream event object.
 * @param {string} type
 * @param {object} [data]
 * @returns {{ type: string, data?: object }}
 */
export function event(type, data) {
  return data ? { type, data } : { type };
}

/**
 * Extract text from a content block (handles string or array-of-parts).
 * @param {string|Array} content
 * @returns {string}
 */
export function textFromContent(content) {
  if (typeof content === "string") return content;
  if (!Array.isArray(content)) return "";
  return content.map((part) => part.text || part.input_text || "").join("");
}

/**
 * Parse a value that may be a JSON string or already an object.
 * @param {any} value
 * @returns {object}
 */
export function parseArgs(value) {
  if (!value) return {};
  if (typeof value === "object") return value;
  try { return JSON.parse(value); } catch { return {}; }
}

/**
 * Build HTTP headers for a provider request, applying resolved credentials.
 *
 * Strategy selection:
 *   - "x-api-key": writes resolved.credential into x-api-key header
 *   - "bearer":    writes "Bearer {credential}" into Authorization
 *   - "key-query": appends ?key= to the URL (Gemini API key style)
 *
 * @param {Record<string,string>} baseHeaders
 * @param {object} resolved       – resolved model config from the host
 * @param {("x-api-key"|"bearer"|"key-query")} [authStrategy="bearer"]
 * @returns {Record<string,string>}
 */
export function headerMap(baseHeaders, resolved, authStrategy) {
  const strategy = authStrategy || "bearer";
  const headers = { ...baseHeaders, ...(resolved.resolvedHttpHeaders || {}) };
  const hasAuth = Object.keys(headers).some((key) =>
    ["x-api-key", "authorization"].includes(key.toLowerCase())
  );
  if (hasAuth || !resolved.credential) return headers;
  if (strategy === "x-api-key") {
    headers["x-api-key"] = resolved.credential;
  } else {
    headers.Authorization = `Bearer ${resolved.credential}`;
  }
  return headers;
}

/**
 * Extract normalized usage from a provider response chunk.
 *
 * Supports these upstream shapes:
 *   - Anthropic:    { message: { usage: { input_tokens, output_tokens } } }
 *   - Chat Completions: { usage: { prompt_tokens, completion_tokens } }
 *   - Gemini:      { usageMetadata: { promptTokenCount, candidatesTokenCount } }
 *   - Responses:   { response: { usage: { input_tokens, output_tokens } } }
 *
 * @param {object} value         – raw upstream chunk
 * @param {object} [previous]    – previous usage (for incremental updates)
 * @returns {{ promptTokens: number, completionTokens: number, totalTokens: number } | null}
 */
export function usageFrom(value, previous) {
  if (!value) return null;
  var raw = (value.message && value.message.usage) || value.usage;
  if (!raw && value.response && value.response.usage) raw = value.response.usage;
  if (!raw && value.usageMetadata) {
    raw = {
      input_tokens: value.usageMetadata.promptTokenCount || 0,
      output_tokens: value.usageMetadata.candidatesTokenCount || 0,
    };
  }
  if (!raw) return null;

  var pt = raw.input_tokens || raw.prompt_tokens || raw.inputTokens || 0;
  pt = pt + (raw.cache_creation_input_tokens || 0) + (raw.cache_read_input_tokens || 0);
  if (!pt && previous && previous.promptTokens) pt = previous.promptTokens;

  var ct = raw.output_tokens || raw.completion_tokens || raw.outputTokens || 0;
  if (!ct && previous && previous.completionTokens) ct = previous.completionTokens;

  var tt = raw.total_tokens || raw.totalTokens || (pt + ct);
  if (!pt && !ct && !tt) return null;
  return { promptTokens: pt, completionTokens: ct, totalTokens: tt };
}

/**
 * Apply a shared subset of ModelRequestConfig to a mutable payload object.
 *
 * @param {object} payload     – mutated in place
 * @param {object|null} cfg    – resolved.requestConfig
 * @param {string[]} [skip]    – keys to skip (e.g. provider-specific fields)
 */
export function applyRequestConfig(payload, cfg, skip) {
  cfg = cfg || {};
  const excluded = new Set(skip || []);
  if (cfg.temperature != null && !excluded.has("temperature")) payload.temperature = cfg.temperature;
  if (cfg.topP != null && !excluded.has("topP")) payload.top_p = cfg.topP;
  if (cfg.topK != null && !excluded.has("topK")) payload.top_k = cfg.topK;
  if (cfg.maxOutputTokens != null && !excluded.has("maxOutputTokens")) payload.max_tokens = cfg.maxOutputTokens;
  if (cfg.frequencyPenalty != null && !excluded.has("frequencyPenalty")) payload.frequency_penalty = cfg.frequencyPenalty;
  if (cfg.presencePenalty != null && !excluded.has("presencePenalty")) payload.presence_penalty = cfg.presencePenalty;
  for (const [key, value] of Object.entries(cfg.extraBody || {})) {
    const normalized = key.trim().toLowerCase();
    if (normalized && !["messages", "model", "stream"].includes(normalized)) {
      payload[key] = value;
    }
  }
}

File: src-tauri/src/plugin_runtime/sdk.rs
//! Shared plugin SDK modules provided via deno_core's `StaticModuleLoader`.
//!
//! Built-in and user plugins can `import` from these pre-registered modules
//! instead of bundling duplicate helper code. The SDK sources are compiled
//! into the binary, so no filesystem access is needed at runtime.

use deno_core::{StaticModuleLoader, url::Url};
use std::rc::Rc;

/// Module specifier for the AgentJax SDK shared library.
pub const SDK_MODULE_SPECIFIER: &str = "builtin:/agentjax/sdk";

/// Return an `Rc<dyn ModuleLoader>` pre-populated with all built-in SDK modules.
///
/// Call this once when initialising the plugin runtime and clone the `Rc` for
/// each new plugin `JsRuntime` instance so all plugins share the same module
/// resolution table.
pub fn create_sdk_module_loader() -> Rc<dyn deno_core::ModuleLoader> {
    let sdk_url = Url::parse(SDK_MODULE_SPECIFIER).expect("SDK module specifier is a valid URL");
    let sdk_source: &'static str = include_str!("../../builtin-plugins/sdk/sdk.js");
    let loader = StaticModuleLoader::new([(sdk_url, sdk_source)]);
    Rc::new(loader)
}

#[cfg(test)]
mod tests {
    use super::*;
    use deno_core::{JsRuntime, RuntimeOptions};

    #[test]
    fn sdk_module_is_loadable() {
        let module_loader = create_sdk_module_loader();
        let mut runtime = JsRuntime::new(RuntimeOptions {
            module_loader: Some(module_loader),
            ..Default::default()
        });

        let result = runtime.execute_script(
            "<test-sdk-import>",
            r#"
(async () => {
  const sdk = await import("builtin:/agentjax/sdk");
  if (typeof sdk.withQuery !== "function") throw new Error("withQuery missing");
  if (typeof sdk.event !== "function") throw new Error("event missing");
  if (typeof sdk.textFromContent !== "function") throw new Error("textFromContent missing");
  if (typeof sdk.parseArgs !== "function") throw new Error("parseArgs missing");
  if (typeof sdk.headerMap !== "function") throw new Error("headerMap missing");
  if (typeof sdk.usageFrom !== "function") throw new Error("usageFrom missing");
  if (typeof sdk.applyRequestConfig !== "function") throw new Error("applyRequestConfig missing");
  const result = sdk.withQuery("https://example.com/path", { foo: "bar", baz: "qux" });
  if (result !== "https://example.com/path?foo=bar&baz=qux") {
    throw new Error("withQuery returned unexpected: " + result);
  }
  return "ok";
})()
"#,
        );
        assert!(result.is_ok(), "SDK module should load and export all helpers");
    }

    #[test]
    fn sdk_header_map_bearer_strategy() {
        let module_loader = create_sdk_module_loader();
        let mut runtime = JsRuntime::new(RuntimeOptions {
            module_loader: Some(module_loader),
            ..Default::default()
        });

        let result = runtime.execute_script(
            "<test-header-map>",
            r#"
(async () => {
  const sdk = await import("builtin:/agentjax/sdk");
  const headers = sdk.headerMap(
    { "Content-Type": "application/json" },
    { credential: "sk-test", resolvedHttpHeaders: {} },
    "bearer"
  );
  if (headers.Authorization !== "Bearer sk-test") throw new Error("bearer auth failed");
  if (headers["Content-Type"] !== "application/json") throw new Error("base headers lost");
  return "ok";
})()
"#,
        );
        assert!(result.is_ok(), "headerMap bearer strategy should work");
    }

    #[test]
    fn sdk_usage_from_handles_all_shapes() {
        let module_loader = create_sdk_module_loader();
        let mut runtime = JsRuntime::new(RuntimeOptions {
            module_loader: Some(module_loader),
            ..Default::default()
        });

        // Anthropic shape
        let r1 = runtime.execute_script(
            "<test-usage-anthropic>",
            r#"
(async () => {
  const sdk = await import("builtin:/agentjax/sdk");
  const u = sdk.usageFrom({ message: { usage: { input_tokens: 10, output_tokens: 20 } } });
  if (!u || u.promptTokens !== 10 || u.completionTokens !== 20) throw new Error("anthropic");
  return "ok";
})()
"#,
        );
        assert!(r1.is_ok(), "Anthropic usage shape");

        // Chat Completions shape
        let r2 = runtime.execute_script(
            "<test-usage-chat>",
            r#"
(async () => {
  const sdk = await import("builtin:/agentjax/sdk");
  const u = sdk.usageFrom({ usage: { prompt_tokens: 5, completion_tokens: 15 } });
  if (!u || u.promptTokens !== 5 || u.completionTokens !== 15) throw new Error("chat");
  return "ok";
})()
"#,
        );
        assert!(r2.is_ok(), "Chat Completions usage shape");

        // Gemini shape
        let r3 = runtime.execute_script(
            "<test-usage-gemini>",
            r#"
(async () => {
  const sdk = await import("builtin:/agentjax/sdk");
  const u = sdk.usageFrom({ usageMetadata: { promptTokenCount: 8, candidatesTokenCount: 12 } });
  if (!u || u.promptTokens !== 8 || u.completionTokens !== 12) throw new Error("gemini");
  return "ok";
})()
"#,
        );
        assert!(r3.is_ok(), "Gemini usage shape");
    }

    #[test]
    fn sdk_event_with_and_without_data() {
        let module_loader = create_sdk_module_loader();
        let mut runtime = JsRuntime::new(RuntimeOptions {
            module_loader: Some(module_loader),
            ..Default::default()
        });

        let result = runtime.execute_script(
            "<test-event>",
            r#"
(async () => {
  const sdk = await import("builtin:/agentjax/sdk");
  const e1 = sdk.event("OutputTextStarted");
  if (e1.type !== "OutputTextStarted" || e1.data !== undefined) throw new Error("no-data");
  const e2 = sdk.event("ToolCallStarted", { call_id: "call_1" });
  if (e2.type !== "ToolCallStarted" || e2.data.call_id !== "call_1") throw new Error("with-data");
  return "ok";
})()
"#,
        );
        assert!(result.is_ok(), "event helper should work");
    }
}
